streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,639 @@
+                sleep(0.07)
+        elif state == _ShardState.DOWNLOADED:
+            lock.release()
+        else:
+            raise RuntimeError('Unknown shard state')
+
+    def _download_thread(self, epoch: int, sample_ids: NDArray[np.int64]) -> None:
+        """Download the relevant shards in the background while we are being iterated.
+
+        Args:
+            epoch (int): Which epoch. On noticing that a new epoch has started, we exit this thread
+                because a new one will soon be running, with different sample IDs.
+            sample_ids (NDArray[np.int64]): The samples to download the shards of.
+        """
+        # Create or attach shard_states array.
+        name = f'{self._prefix}_shard_states'
+        size = len(self.shard_sizes) * np.uint8(0).nbytes
+        try:
+            shm = SharedMemory(name, True, size)
+        except:
+            shm = SharedMemory(name)
+        shard_states = np.ndarray(len(self.shard_sizes), buffer=shm.buf, dtype=np.uint8)
+
+        filename = os.path.join('/tmp', 'mds', self._prefix, '_shard_states_filelock')
+        shard_states_lock = FileLock(filename)
+
+        # Download loop.
+        num_samples = len(sample_ids)
+        while True:
+            # If we've started a new epoch early (__iter__ was called again), exit this thread
+            # because there can only be one epoch at once.
+            if epoch != self.next_epoch - 1:
+                break
+
+            # If we're out of samples this epoch, exit this thread because we are done downloading.
+            if self._download_index == num_samples:
+                break
+
+            # If we are requested to only pre-download so many samples, if we have as many or more
+            # downloaded already, we wait and check again later.
+            if self.predownload is not None:
+                samples_ahead = self._download_index - self._iter_index
+                if self.predownload <= samples_ahead:
+                    sleep(0.07)
+                    continue
+
+            # Download and decompress the shard for this sample, if not already done.
+            sample_id = sample_ids[self._download_index]
+            shard_id, _ = self.index.find_sample(sample_id)
+            self._download_or_await_shard(shard_states_lock, shard_states, shard_id)
+            self._download_index += 1
+
+    def __iter__(self) -> Iterator[Dict[str, Any]]:
+        """Iterate over all the samples in our partition.
+
+        Returns:
+            Iterator[Dict[str, Any]]: Each sample.
+        """
+        # Load resume state and create the new training session.
+        presumed_epoch = self.next_epoch
+        epoch, old_sessions = self._resume(presumed_epoch)
+        cur_session, _ = self._create_cur_session(epoch)
+
+        world = World()
+        assert world.num_workers == self._barrier.count
+        self._barrier()
+
+        # Update the pre-incremented epoch counter.
+        world = World()
```

Comment:
Do we need to instantiate the `World()` again here? Since like it is already instantiated 3 lines above!!

Reply:
fixed